test/python/transpiler/test_elide_permutations.py

```diff
@@ -0,0 +1,348 @@
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 0)
+        expected.barrier(0, 1, 2)
+        expected.measure(2, 0)
+        expected.measure(1, 1)
+        expected.measure(0, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_permutation_at_beginning(self):
+        """Test swap in beginning of bell is elided."""
+        qc = QuantumCircuit(3, 3)
+        qc.append(PermutationGate([2, 1, 0]), [0, 1, 2])
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(2)
+        expected.cx(1, 0)
+        expected.barrier(0, 1, 2)
+        expected.measure(2, 0)
+        expected.measure(1, 1)
+        expected.measure(0, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_permutation_at_end(self):
+        """Test swap in beginning of bell is elided."""
+        qc = QuantumCircuit(3, 3)
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+        qc.append(PermutationGate([2, 1, 0]), [0, 1, 2])
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 2)
+        expected.barrier(0, 1, 2)
+        expected.measure(0, 0)
+        expected.measure(1, 1)
+        expected.measure(2, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_swap_and_permutation(self):
+        """Test a combination of swap and permutation gates."""
+        qc = QuantumCircuit(3, 3)
+        qc.append(PermutationGate([2, 1, 0]), [0, 1, 2])
+        qc.swap(0, 2)
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 2)
+        expected.barrier(0, 1, 2)
+        expected.measure(0, 0)
+        expected.measure(1, 1)
+        expected.measure(2, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_permutation_before_measure(self):
+        """Test swap in beginning of bell is elided."""
+        qc = QuantumCircuit(3, 3)
+        qc.h(0)
+        qc.cx(1, 2)
+        qc.barrier(0, 1, 2)
+        qc.append(PermutationGate([1, 2, 0]), [0, 1, 2])
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+        qc.measure(2, 2)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.cx(1, 2)
+        expected.barrier(0, 1, 2)
+        expected.measure(1, 0)
+        expected.measure(2, 1)
+        expected.measure(0, 2)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
```

Comment:
Some other tests that might be good:

- Test a circuit that only contains swaps and permutations, and transpile it.
- Check unitary equivalence for a circuit with swaps and permutations that are elided.

Reply:
It's a good thing you asked me to add these tests, the unitary equivalence with a mix of swap and permutations caused the failure I was expecting in `test_unitary_equivalence_with_elide_and_routing`. I don't think I'll have enough time to include this in 0.45.0 so I'm going to defer it to the next release.